diff --git a/src/cryptonote_config.h b/src/cryptonote_config.h
index 82891b9de..10aaf94a6 100644
--- a/src/cryptonote_config.h
+++ b/src/cryptonote_config.h
@@ -133,7 +133,7 @@
 #define DEFAULT_RPC_SOFT_LIMIT_SIZE                     25 * 1024 * 1024 // 25 MiB
 #define MAX_RPC_CONTENT_LENGTH                          1048576 // 1 MB
 
-#define P2P_LOCAL_WHITE_PEERLIST_LIMIT                  1000
+#define P2P_LOCAL_WHITE_PEERLIST_LIMIT                  10000
 #define P2P_LOCAL_GRAY_PEERLIST_LIMIT                   5000
 
 #define P2P_DEFAULT_CONNECTIONS_COUNT                   12
diff --git a/src/p2p/net_node.h b/src/p2p/net_node.h
index d7cbaa567..d7b89017f 100644
--- a/src/p2p/net_node.h
+++ b/src/p2p/net_node.h
@@ -371,7 +371,6 @@ namespace nodetool
     bool make_new_connection_from_anchor_peerlist(const std::vector<anchor_peerlist_entry>& anchor_peerlist);
     bool make_new_connection_from_peerlist(network_zone& zone, bool use_white_list);
     bool try_to_connect_and_handshake_with_new_peer(const epee::net_utils::network_address& na, bool just_take_peerlist = false, uint64_t last_seen_stamp = 0, PeerType peer_type = white, uint64_t first_seen_stamp = 0);
-    size_t get_random_index_with_fixed_probability(size_t max_index);
     bool is_peer_used(const peerlist_entry& peer);
     bool is_peer_used(const anchor_peerlist_entry& peer);
     bool is_addr_connected(const epee::net_utils::network_address& peer);
diff --git a/src/p2p/net_node.inl b/src/p2p/net_node.inl
index 91a54f43e..c15256f09 100644
--- a/src/p2p/net_node.inl
+++ b/src/p2p/net_node.inl
@@ -1286,19 +1286,6 @@ namespace nodetool
   }
   //-----------------------------------------------------------------------------------
   template<class t_payload_net_handler>
-  size_t node_server<t_payload_net_handler>::get_random_index_with_fixed_probability(size_t max_index)
-  {
-    //divide by zero workaround
-    if(!max_index)
-      return 0;
-
-    size_t x = crypto::rand<size_t>()%(16*max_index+1);
-    size_t res = (x*x*x)/(max_index*max_index*16*16*16); //parabola \/
-    MDEBUG("Random connection index=" << res << "(x="<< x << ", max_index=" << max_index << ")");
-    return res;
-  }
-  //-----------------------------------------------------------------------------------
-  template<class t_payload_net_handler>
   bool node_server<t_payload_net_handler>::is_peer_used(const peerlist_entry& peer)
   {
     const auto zone = peer.adr.get_zone();
@@ -1718,7 +1705,7 @@ namespace nodetool
         // else, for step 1 / second pass of inner try loop, take all peers from all subnets
 
         // Take as many candidates as we need and care about stripes if pruning
-        const size_t limit = use_white_list ? 20 : std::numeric_limits<size_t>::max();
+        const size_t limit = std::numeric_limits<size_t>::max();
         for (const peerlist_entry &peer : candidate_peers) {
           if (filtered.size() >= limit)
             break;
@@ -1746,9 +1733,8 @@ namespace nodetool
       size_t random_index;
       if (use_white_list)
       {
-        // If using the white list, we first pick in the set of peers we've already been using earlier;
-        // that "fixed probability" heavily favors the peers most recently seen in the candidate list
-        random_index = get_random_index_with_fixed_probability(filtered.size() - 1);
+        random_index = crypto::rand_idx(filtered.size());
+        CHECK_AND_ASSERT_MES(random_index < filtered.size(), false, "random_index < filtered.size() failed!!");
 
         CRITICAL_REGION_LOCAL(m_used_stripe_peers_mutex);
         if (next_needed_pruning_stripe > 0 && next_needed_pruning_stripe <= (1ul << CRYPTONOTE_PRUNING_LOG_STRIPES) && !m_used_stripe_peers[next_needed_pruning_stripe-1].empty())
@@ -1935,12 +1921,12 @@ namespace nodetool
             && make_expected_connections_count(zone.second, gray, zone.second.m_config.m_net_config.max_out_connection_count));
         }else
         {
-          //start from grey list
-          while (get_outgoing_connections_count(zone.second) < zone.second.m_config.m_net_config.max_out_connection_count
-            && make_expected_connections_count(zone.second, gray, zone.second.m_config.m_net_config.max_out_connection_count));
-          //and then do white list
+          //start from white list
           while (get_outgoing_connections_count(zone.second) < zone.second.m_config.m_net_config.max_out_connection_count
             && make_expected_connections_count(zone.second, white, zone.second.m_config.m_net_config.max_out_connection_count));
+          //and then do grey list
+          while (get_outgoing_connections_count(zone.second) < zone.second.m_config.m_net_config.max_out_connection_count
+            && make_expected_connections_count(zone.second, gray, zone.second.m_config.m_net_config.max_out_connection_count));
         }
         if(zone.second.m_net_server.is_stop_signal_sent())
           return false;
